### DISCUSSION TITLE
How to enable and save API Load Test Results Raw data

### DISCUSSION THREAD

**Discussion by rrnareshkumar** [artilleryio/artillery#2118] [Artillery]
Hi , Please suggest how to enable and save API Load Test Results Raw data.

For example , when load test is executed and 1000 requests are triggered during load test , would like to enable and save raw data to view details such as response time of each and every request and its corresponding timestamp , sent bytes , received bytes etc . something like below example . 

timeStamp | response   time | label | responseCode | responseMessage | failureMessage | received bytes | sentBytes | API URL
1693989067152 | 2231 | Test API 2 | 200 | OK |   | 541 | 129 |  example1.com
1693989069388 | 1794 | Test API 1 | 200 | OK |   | 1388 | 122 |  example2.com
1693989071185 | 298 | Test API 2 | 200 | OK |   | 541 | 129 |  example1.com
1693989069651 | 1832 | Test API 2 | 200 | OK |   | 541 | 129 |  example2.com

**Comment by InesNi** [comment]
Hello @rrnareshkumar 👋🏻

Artillery collects data in 10s intervals into intermediate reports which aggregate the results for that time window with relevant statistical summaries. This is common for load testing tools, given the sheer amount of data they can produce.

Make sure to also enable the [metrics-by-endpoint](https://www.artillery.io/docs/reference/extensions/metrics-by-endpoint) plugin to get per-endpoint information.

If you want to collect data for each request, the best bet would be to use one of the tracing options from the `publish-metrics` plugin (https://www.artillery.io/docs/reference/extensions/publish-metrics). We currently support `Honeycomb` and `Lightstep` as targets for tracing, and we are in the process of releasing OpenTelemetry support which will allow you to send traces to more platforms (any that support ingestion via OTLP).

Hope that helps! 🚀

**Reply by rrnareshkumar** [reply]
Hi , thanks for the response . 

yes , I want to collect data for each request. I am executing load test from my local machine and don't have access to artillery supported targets(APM or monitoring tools ) such as Honeycomb , Lightstep , AWS cloud watch , data dog , new relic etc  .

I would like to understand without other APM or monitoring tools dependencies , is it possible for me to enable and collect this raw data  (data for each request) and save it locally (for example in .csv file) to perform performance results analysis ? Please suggest .

**Reply by InesNi** [reply]
Hi @rrnareshkumar,

The other alternative I see is you could take advantage of Artillery's **hooks** and run some custom JS logic at certain points in scenario execution like specified in the [documentation](https://www.artillery.io/docs/reference/engines/http#writing-custom-logic-in-javascript). For instance `beforeRequest` and `afterRequest` hooks enable you to interact directly with each request and response object and to extract what you need in which ever form you need. You can even create custom metrics like specified in the documentation [here](https://www.artillery.io/docs/reference/extension-apis#tracking-custom-metrics).

However, logging every single request like this may have a performance implication, due to the additional javascript being executed. It would be recommended to use Artillery's own data aggregation.

Hope this helps!
